src/trio/testing/_raises_group.py

```diff
@@ -0,0 +1,305 @@
+    from exceptiongroup import BaseExceptionGroup
+
+E = TypeVar("E", bound=BaseException)
+
+
+# minimal version of pytest.ExceptionInfo in case it is not available
+@final
+class _ExceptionInfo(Generic[E]):
+    _excinfo: tuple[type[E], E, types.TracebackType] | None
+
+    def __init__(self, excinfo: tuple[type[E], E, types.TracebackType] | None):
+        self._excinfo = excinfo
+
+    def fill_unfilled(self, exc_info: tuple[type[E], E, types.TracebackType]) -> None:
+        """Fill an unfilled ExceptionInfo created with ``for_later()``."""
+        assert self._excinfo is None, "ExceptionInfo was already filled"
+        self._excinfo = exc_info
+
+    @classmethod
+    def for_later(cls) -> _ExceptionInfo[E]:
+        """Return an unfilled ExceptionInfo."""
+        return cls(None)
+
+    @property
+    def type(self) -> type[E]:
+        """The exception class."""
+        assert (
+            self._excinfo is not None
+        ), ".type can only be used after the context manager exits"
+        return self._excinfo[0]
+
+    @property
+    def value(self) -> E:
+        """The exception value."""
+        assert (
+            self._excinfo is not None
+        ), ".value can only be used after the context manager exits"
+        return self._excinfo[1]
+
+    @property
+    def tb(self) -> types.TracebackType:
+        """The exception raw traceback."""
+        assert (
+            self._excinfo is not None
+        ), ".tb can only be used after the context manager exits"
+        return self._excinfo[2]
+
+
+# this may bite users with type checkers not using pytest, but that should
+# be rare and give quite obvious errors in tests trying to do so.
+# Ignoring "incorrect import of pytest", it makes sense in this situation.
+if TYPE_CHECKING:
+    from pytest import ExceptionInfo  # noqa: PT013
+
+else:
+    try:
+        from pytest import ExceptionInfo  # noqa: PT013
+    except ImportError:  # pragma: no cover
+        ExceptionInfo = _ExceptionInfo
+
+
+# copied from pytest.ExceptionInfo
+def _stringify_exception(exc: BaseException) -> str:
+    return "\n".join(
+        [
+            str(exc),
+            *getattr(exc, "__notes__", []),
+        ]
+    )
+
+
+@final
+class Matcher(Generic[E]):
+    def __init__(
+        self,
+        exception_type: type[E] | None = None,
+        match: str | Pattern[str] | None = None,
+        check: Callable[[E], bool] | None = None,
+    ):
+        if exception_type is None and match is None and check is None:
+            raise ValueError("You must specify at least one parameter to match on.")
+        if exception_type is not None and not issubclass(exception_type, BaseException):
+            raise ValueError(
+                f"exception_type {exception_type} must be a subclass of BaseException"
+            )
+        self.exception_type = exception_type
+        self.match = match
+        self.check = check
+
+    def matches(self, exception: E) -> TypeGuard[E]:
```

Comment:
Shouldn't this be `exception: BaseException` to allow this to be used to check something that may not match?

Reply:
Type guards are special functions that take an input that could be a union of lots of things and return a boolean. If the value is True, the input is supposed to be confirmed to be of the type given as a parameter of the type guard. I think I agree with TeamSpen210 here, this type guard doesn't make a lot of sense just by looking at the snippet shown here, because it's as if we are just confirming that `E` is indeed `E`, but if it wasn't then type checkers should raise a `invalid parameter value` or something error when we try to call this function.

Reply:
ah, that caught `check` being off as well - instead of `Callable[[E], bool]` it's now `Callable[[BaseException], TypeGuard[E]]`

Reply:
I don't think you want to make it use typeguard, since it's unlikely people will be using that annotation. Not sure if checkers will infer that for lambdas also. Actually this probably needs an overload - if `exception_type` is supplied it's `Callable[[E], bool]` (so you can check that exception's attrs), but if not it would be `Callable[[BaseException], bool]`. 

Best would be to type the attribute as `Callable[[E], bool]`, do the overloads I mentioned below, and in the case that `exception_type` isn't supplied type `self: Match[BaseException]`. You'll need a cast if you leave `exception_type` able to be None I think.

Reply:
Forcing everybody to use `TypeGuard` would be amusing, but probably wiser not to.